File: baekjoon/bj1916.java
//Dijkstra_Java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class bj1916 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int city = Integer.parseInt(br.readLine());
        int bus = Integer.parseInt(br.readLine());

        ArrayList<int[]> map[] = new ArrayList[city + 1];
        for (int i = 1; i < city + 1; i++) {
            map[i] = new ArrayList<>();
        }

        StringTokenizer st;
        for (int i = 0; i < bus; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            int pay = Integer.parseInt(st.nextToken());

            map[start].add(new int[] { end, pay });
        }

        st = new StringTokenizer(br.readLine(), " ");
        int go = Integer.parseInt(st.nextToken());
        int finish = Integer.parseInt(st.nextToken());

        System.out.println(serch(map, go, finish));
    }

    private static int serch(ArrayList<int[]>[] map, int go, int finish) {
        int disPay[] = new int[map.length];
        Arrays.fill(disPay, Integer.MAX_VALUE);

        Queue<int[]> que = new LinkedList<>();
        que.add(new int[] { go, 0 });
        disPay[go] = 0;

        while (!que.isEmpty()) {
            int temp[] = que.poll();
            int now = temp[0];
            int pay = temp[1];

            if (disPay[now] < pay)
                continue;

            for (int i[] : map[now]) {
                int next = i[0];
                int nextPay = i[1] + pay;

                if (disPay[next] > nextPay) {
                    que.add(new int[] { next, nextPay });
                    disPay[next] = Math.min(disPay[next], nextPay);
                }
            }

        }
        return disPay[finish];
    }
}

File: baekjoon/bj1916.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val node = br.readLine().toInt()
    val edge = br.readLine().toInt()

    val map = Array(node + 1) { mutableListOf<Pair<Int, Long>>() }
    repeat(edge) {
        val (s, f, cost) = br.readLine().split(" ").map { it.toInt() }
        map[s].add(Pair(f, cost.toLong()))
    }
    val (start, end) = br.readLine().split(" ").map { it.toInt() }




    println(findWay(start, map)[end])


}

private fun findWay(
    start: Int,
    map: Array<MutableList<Pair<Int, Long>>>
): Array<Long> {
    val result = Array(map.size) { Long.MAX_VALUE }
    val pq = PriorityQueue(compareBy<Pair<Int, Long>> { it.second })
    pq.add(Pair(start, 0))
    result[start] = 0

    while (pq.isNotEmpty()) {
        val (cur, curCost) = pq.poll()

        if (curCost > result[cur]) continue

        for ((next, cost) in map[cur]) {
            if (result[next] > cost + curCost) {
                pq.add(Pair(next, cost + curCost))
                result[next] = cost + curCost
            }
        }
    }
    return result
}